CMakeLists.txt

```diff
@@ -207,22 +207,43 @@ endif()
 
 # If available and wanted, use compiler cache.
 if(LIBCELLML_COMPILER_CACHE)
-  if(WIN32)
-    if(CLCACHE_EXE)
-      set(CLCACHEWRAPPER ${CMAKE_CURRENT_BINARY_DIR}/clcachewrapper)
-      execute_process(COMMAND ${CMAKE_C_COMPILER} /O2 /Fe${CLCACHEWRAPPER} ${CMAKE_CURRENT_SOURCE_DIR}/cmake/clcachewrapper.c
-                      RESULT_VARIABLE RESULT OUTPUT_QUIET ERROR_QUIET)
-      if(RESULT EQUAL 0)
-        set(CMAKE_CXX_COMPILER_LAUNCHER ${CLCACHEWRAPPER})
-      endif()
-    endif()
+  if(BUILDCACHE_EXE)
+    set(CMAKE_CXX_COMPILER_LAUNCHER ${BUILDCACHE_EXE})
   else()
-    if(CCACHE_EXE)
-      set(CMAKE_CXX_COMPILER_LAUNCHER ${CCACHE_EXE})
+    if(WIN32)
+      if(CLCACHE_EXE)
+        set(CLCACHEWRAPPER ${CMAKE_CURRENT_BINARY_DIR}/clcachewrapper)
+        execute_process(COMMAND ${CMAKE_C_COMPILER} /O2 /Fe${CLCACHEWRAPPER} ${CMAKE_CURRENT_SOURCE_DIR}/cmake/clcachewrapper.c
+                        RESULT_VARIABLE RESULT OUTPUT_QUIET ERROR_QUIET)
+        if(RESULT EQUAL 0)
+          set(CMAKE_CXX_COMPILER_LAUNCHER ${CLCACHEWRAPPER})
+        endif()
+      endif()
+    else()
+      if(CCACHE_EXE)
+        set(CMAKE_CXX_COMPILER_LAUNCHER ${CCACHE_EXE})
+      endif()
     endif()
   endif()
 endif()
 
+# On Windows, if we are using MSVC then replace some compiler flags.
+if(WIN32 AND MSVC)
+  # If we are treating warnings as errors then remove the /W3 option since it
+  # will eventually be replaced with the /W4 and /WX options (see
+  # target_warnings_as_errors()).
+  if(LIBCELLML_TREAT_WARNINGS_AS_ERRORS)
+    replace_compiler_flag("/W3" "")
+  endif()
+
+  # Replace the /Zi option (i.e. enable debugging information), if present, with
+  # the /Z7 option (i.e. enable old-style debugging information) otherwise
+  # caching (be it with buildcache or clcache) won't work.
+  if(BUILDCACHE_EXE OR CLCACHE_EXE)
+    replace_compiler_flag("/Zi" "/Z7")
+  endif()
+endif()
```

Comment:
All of this should be pushed into `common.cmake`.

Reply:
Why? Wouldn't you put in `common.cmake` functions/macros that are called in various places? The above code is only ever executed once, so we might as well leave it here unless you want to keep `CMakeLists.txt` as short as possible... and make maintenance / reading the code unnecessarily "difficult" by forcing people to look for things elsewhere...